internal/compiler/langtype.rs

```diff
@@ -234,10 +225,167 @@ impl Type {
+    /// otherwise, returns None
+    pub fn default_unit(&self) -> Option<Unit> {
+        match self {
+            Type::Duration => Some(Unit::Ms),
+            Type::PhysicalLength => Some(Unit::Phx),
+            Type::LogicalLength => Some(Unit::Px),
+            // Unit::Percent is special that it does not combine with other units like
+            Type::Percent => None,
+            Type::Angle => Some(Unit::Deg),
+            Type::Invalid => None,
+            Type::Void => None,
+            Type::InferredProperty | Type::InferredCallback => None,
+            Type::Callback { .. } => None,
+            Type::Function { .. } => None,
+            Type::Float32 => None,
+            Type::Int32 => None,
+            Type::String => None,
+            Type::Color => None,
+            Type::Image => None,
+            Type::Bool => None,
+            Type::Model => None,
+            Type::PathData => None,
+            Type::Easing => None,
+            Type::Brush => None,
+            Type::Array(_) => None,
+            Type::Struct { .. } => None,
+            Type::Enumeration(_) => None,
+            Type::UnitProduct(_) => None,
+            Type::ElementReference => None,
+            Type::LayoutCache => None,
+        }
+    }
+
+    /// Return a unit product vector even for single scalar
+    pub fn as_unit_product(&self) -> Option<Vec<(Unit, i8)>> {
+        match self {
+            Type::UnitProduct(u) => Some(u.clone()),
+            Type::Float32 | Type::Int32 => Some(Vec::new()),
+            _ => self.default_unit().map(|u| vec![(u, 1)]),
+        }
+    }
+}
+
+impl Default for Type {
+    fn default() -> Self {
+        Self::Invalid
+    }
+}
+
+/// Information about properties in NativeClass
+#[derive(Debug, Clone)]
+pub struct BuiltinPropertyInfo {
+    /// The property type
+    pub ty: Type,
+    /// When set, this is the initial value that we will have to set if no other binding were specified
+    pub default_value: Option<Expression>,
+    /// Most properties are just set from the .slint code and never modified by the native code.
+    /// But some properties, such as `TouchArea::pressed` are being set by the native code, these
+    /// are output properties which are meant to be read by the .slint.
+    /// `is_native_output` is true if the native item can modify the property.
+    pub is_native_output: bool,
+}
+
+impl BuiltinPropertyInfo {
+    pub fn new(ty: Type) -> Self {
+        Self { ty, default_value: None, is_native_output: false }
+    }
+}
+
+/// The base of an element
+#[derive(Clone, Debug)]
+pub enum ElementType {
+    /// The element is based of a component
+    Component(Rc<Component>),
+    /// The element is a builtin element
+    Builtin(Rc<BuiltinElement>),
+    /// The native type was resolved by the resolve_native_class pass.
+    Native(Rc<NativeClass>),
+    /// The base element couldn't be looked up
+    Error,
```

Comment:
Perhaps `Invalid` would be a better name to express what is essential the invalid state?

Reply:
It is a valid state, just one caused by an error in the .slint file.